### ISSUE TITLE
The ONNX inference results for metric depth are very poor.

### ISSUE THREAD

**Issue by dnflslwlq** [DepthAnything/Depth-Anything-V2#49]
Hello @LiheYoung ,

I am trying to use the metric depth estimation pretrained model (vkitti) from Depth Anything V2 to convert it to ONNX and perform inference. (test : kitti dataset)

The relative depth did not show significant performance degradation. However, when converting the metric depth to ONNX and performing inference, the performance degradation was very severe.

I tried converting it with Dynamo and using a general script, but the result was the same.

What could be the cause of this?

The backbone model structures for relative and metric are not different, so why does the performance significantly degrade when converting to ONNX? I conducted the conversion using FP32. (not FP16)

The following results are from relative ONNX and metric ONNX, respectively.

![0000000000](https://github.com/DepthAnything/Depth-Anything-V2/assets/114914559/0241b68d-6de4-4576-be34-34f3c06b82b2)
![0000000000](https://github.com/DepthAnything/Depth-Anything-V2/assets/114914559/e6b87bfe-f0b6-4271-b666-b3db9dd71edc)

**Comment by Edric-star** [comment]
Hi, have you tried comparing the performance of these two (relative and metric) without conveting them into ONNX?

**Comment by dnflslwlq** [comment]
@Edric-star 
Yes. As a result of my tests, strangely enough, the Base model showed the best performance in terms of "numerical values" in the Pytorch models. Since this is based on testing with only one sequence (2011_09_26_drive_0009_sync), I am now planning to test it with other sequences as well.

The comparison in Pytorch is subtle, but when tested after converting to TensorRT, the difference becomes more pronounced.

**Comment by dnflslwlq** [comment]
@zjmooon 
Please refer to the following links:
https://github.com/fabio-sim/Depth-Anything-ONNX
https://github.com/spacewalk01/depth-anything-tensorrt

**Comment by dnflslwlq** [comment]
For your reference, I have found a solution to this issue.
When converting to ONNX, you need to set the dummy input to a multiple of 14 that is closest to the "resolution of the test data" rather than 518 518. This ensures minimal degradation, although latency cannot be guaranteed.

In the case of the Kitti datasets, the original resolution is 375 1242, which cannot be converted to ONNX as is. This is because the patch size must be a multiple of 14 due to the model layer characteristics.
Therefore, you should change the ONNX dummy input to 378 1246 and ensure proper pre-processing and post-processing so that the output is in the original size during inference.

**Comment by ze-n-g** [comment]
@dnflslwlq，Hello, would you like to ask whether you use the official open source metric model or the metric model you trained yourself?

**Comment by dnflslwlq** [comment]
hello @ze-n-g , I used the official open-source metric model.

**Comment by lalawayne** [comment]
> I have found a solution to this issue. When converting to ONNX, you need to set the dummy input to a multiple of 14 that is closest to the "resolution of the test data" rather than 518 518. This ensures minimal degradation, although latency cannot be guaranteed.
> 
> In the case of the Kitti datasets, the original resolution is 375 1242, which cannot be converted to ONNX as is. This is because the patch size must be a multiple of 14 due to the model layer characteristics. Therefore, you should change the ONNX dummy input to 378 1246 and ensure proper pre-processing and post-processing so that the output is in the original size during inference.

hi，I encountered the same problem as you，When I used your Depth anything V2 vits.onnx model to test my own data set, there was a problem. The test results were different from the pytorch version of V2 vits.pth. The test results of the pytorch version were significantly better than the onnx model.
Below is the code I used for testing:
Pytorch version:
python run.py --img-path ... --outdir ... --encoder vits
ONNX version（https://github.com/fabio-sim/Depth-Anything-ONNX）:
python dynamo.py export --encoder vits --output weights/vits.onnx --opset 17
python dynamo.py infer weights/vits.onnx -i ... -o ...
Other parameters are default parameters.Could you tell me how to solve this problem? I sincerely hope to get your answer.

**Comment by dnflslwlq** [comment]
> > I have found a solution to this issue. When converting to ONNX, you need to set the dummy input to a multiple of 14 that is closest to the "resolution of the test data" rather than 518 518. This ensures minimal degradation, although latency cannot be guaranteed.
> > In the case of the Kitti datasets, the original resolution is 375 1242, which cannot be converted to ONNX as is. This is because the patch size must be a multiple of 14 due to the model layer characteristics. Therefore, you should change the ONNX dummy input to 378 1246 and ensure proper pre-processing and post-processing so that the output is in the original size during inference.
> 
> hi，I encountered the same problem as you，When I used your Depth anything V2 vits.onnx model to test my own data set, there was a problem. The test results were different from the pytorch version of V2 vits.pth. The test results of the pytorch version were significantly better than the onnx model. Below is the code I used for testing: Pytorch version: python run.py --img-path ... --outdir ... --encoder vits ONNX version（[https://github.com/fabio-sim/Depth-Anything-ONNX）](https://github.com/fabio-sim/Depth-Anything-ONNX%EF%BC%89): python dynamo.py export --encoder vits --output weights/vits.onnx --opset 17 python dynamo.py infer weights/vits.onnx -i ... -o ... Other parameters are default parameters.Could you tell me how to solve this problem? I sincerely hope to get your answer.

Hi @lalawayne ,
Please refer to this link. That is what i had written. The detail solution is there
https://github.com/fabio-sim/Depth-Anything-ONNX/issues/16#issue-2385451090

When I tried `dynamo.py`, it had not good result. because width and height are designed top default=518x518 at `def export()` in that script
```
    height: Annotated[
        int,
        typer.Option(
            "-h",
            "--height",
            min=0,
            help="Height of input image. Set to 0 to mark as dynamic.",
            callback=multiple_of_14,
        ),
    ] = 518,
    width: Annotated[
        int,
        typer.Option(
            "-w",
            "--width",
            min=0,
            help="Width of input image. Set to 0 to mark as dynamic.",
            callback=multiple_of_14,
        ),
    ] = 518,

```

**Comment by ze-n-g** [comment]
@dnflslwlq，Thank you very much for your reply. When I tried the official open source model and my own fune-turning model, the depth of the sky part was not accurately predicted. This problem is consistent with yours, but I only used the author's code and did not use the ONNX code. . Do you have any way to solve the problem of inaccurate sky depth prediction of the metric model?
![2a4fb68fe20c5724bd8556f2ef8099a](https://github.com/user-attachments/assets/0e74b120-8531-4f54-8a7c-008f0a654980)
